roxie/topo/toposerver.cpp

```diff
@@ -142,16 +150,22 @@ void timeoutTopology()
     unsigned now = msTick();
     if (now - lastTimeoutCheck < timeoutCheckInterval)
         return;
-    for (auto it = topology.cbegin(); it != topology.cend(); /* no increment */)
+    for (auto it = topology.begin(); it != topology.end(); /* no increment */)
     {
-        unsigned lastSeen = it->second;
+        unsigned lastSeen = it->second.lastSeen;
         if (now-lastSeen > timeoutHeartbeatInterval)
         {
             if (traceLevel)
             {
                 DBGLOG("No heartbeat for %u ms for %s", now-lastSeen, it->first.c_str());
             }
-            it = topology.erase(it);
+            if (now-lastSeen > removeHeartbeatInterval)
```

Comment:
Clearing the instance means that the clients will be easily able to detect that this server has disappeared, rather than having to detect the fact by comparing the new list of servers it gets with the list of servers it knows about.

Once sufficient time has elapsed that everyone should know it has gone, we remove it entirely so that the size of the information doesn't keep growing if servers are continuously replaced.